cmd/util/ledger/migrations/cadence_values_migration_test.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/onflow/flow-go/model/flow"
+
+	_ "github.com/glebarez/go-sqlite"
+)
+
+const snapshotPath string = "test-data/cadence_values_migration/snapshot"
+
+func TestCadenceValuesMigration(t *testing.T) {
+
+	address, err := common.HexToAddress("01cf0e2f2f715450")
+	require.NoError(t, err)
+
+	// Get the old payloads
+
+	payloads := payloadsFromEmulatorSnapshot(t, snapshotPath)
+
+	// Migrate
+
+	rwf := &testReportWriterFactory{}
+	valueMigration := NewCadenceValueMigrator(rwf)
+
+	err = valueMigration.InitMigration(zerolog.Nop(), nil, 0)
+	require.NoError(t, err)
+
+	newPayloads, err := valueMigration.MigrateAccount(nil, address, payloads)
+	require.NoError(t, err)
+
+	// Assert the migrated payloads
+
+	mr, err := newMigratorRuntime(address, newPayloads)
+	require.NoError(t, err)
+
+	storageMap := mr.Storage.GetStorageMap(address, common.PathDomainStorage.Identifier(), false)
+	require.NotNil(t, storageMap)
+	require.Equal(t, 4, int(storageMap.Count()))
+
+	iterator := storageMap.Iterator(mr.Interpreter)
+
+	fullyEntitledAccountReferenceType := interpreter.ConvertSemaToStaticType(nil, sema.FullyEntitledAccountReferenceType)
+
+	var values []interpreter.Value
+	for key, value := iterator.Next(); key != nil; key, value = iterator.Next() {
+		// skip composite values. e.g: FlowToken etc.
+		if _, isComposite := value.(*interpreter.CompositeValue); isComposite {
+			continue
+		}
+		values = append(values, value)
+	}
+
+	// Order is non-deterministic, so use 'ElementsMatch'.
+	assert.ElementsMatch(
+		t,
+		values,
+		[]interpreter.Value{
+			// Both string values should be in the normalized form.
+			interpreter.NewUnmeteredStringValue("Caf\u00E9"),
+			interpreter.NewUnmeteredStringValue("Caf\u00E9"),
+
+			interpreter.NewUnmeteredTypeValue(fullyEntitledAccountReferenceType),
+		},
+	)
+
+	// Check reporters
+
+	reportWriter := valueMigration.reporter.(*testReportWriter)
+
+	// Order is non-deterministic, so use 'ElementsMatch'.
+	assert.ElementsMatch(
+		t,
+		reportWriter.entries,
+		[]any{
+			cadenceValueMigrationReportEntry{
+				Address: interpreter.AddressPath{
+					Address: address,
+					Path: interpreter.PathValue{
+						Identifier: "string_value_1",
+						Domain:     common.PathDomainStorage,
+					},
+				},
+				Migration: "StringNormalizingMigration",
+			},
+			cadenceValueMigrationReportEntry{
+				Address: interpreter.AddressPath{
+					Address: address,
+					Path: interpreter.PathValue{
+						Identifier: "string_value_2",
+						Domain:     common.PathDomainStorage,
+					},
+				},
+				Migration: "StringNormalizingMigration",
+			},
+			cadenceValueMigrationReportEntry{
+				Address: interpreter.AddressPath{
+					Address: address,
+					Path: interpreter.PathValue{
+						Identifier: "type_value_1",
+						Domain:     common.PathDomainStorage,
+					},
+				},
+				Migration: "AccountTypeMigration",
+			},
+		},
+	)
+}
+
+func payloadsFromEmulatorSnapshot(t *testing.T, snapshotPath string) []*ledger.Payload {
```

Comment:
we might want to add this to a more general migration testing utils file, its very useful!

Reply:
Good idea! added in d8669b10ff3953746eab817188bf15717ab8c42a